<a href="https://colab.research.google.com/github/DATA601-CustSegment-Nov2025-to-Feb2026/New-CustSegment-Project/blob/main/Customer_Segmentation_(Merging_1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Libraries & Load Datasets

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import dask.dataframe as dd


from google.colab import drive

In [2]:
# Mount Google Drive
drive.mount('/content/drive')


Mounted at /content/drive


In [3]:
# Paths to the 8 datasets stored in Google Drive
transaction_file_path = '/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/1. Original Datasets/transaction_data.csv'
demographic_file_path = '/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/1. Original Datasets/hh_demographic.csv'
product_file_path = '/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/1. Original Datasets/product.csv'
causal_data_file_path = '/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/1. Original Datasets/causal_data.csv'
coupon_file_path = '/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/1. Original Datasets/coupon.csv'
coupon_redemption_file_path = '/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/1. Original Datasets/coupon_redempt.csv'
campaign_file_path = '/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/1. Original Datasets/campaign_table.csv'
campaign_desc_file_path = '/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/1. Original Datasets/campaign_desc.csv'

# Read the 8 datasets from their CSV files
transaction_df = dd.read_csv(
    transaction_file_path,
    dtype={
        'COUPON_DISC': 'float64',
        'COUPON_MATCH_DISC': 'float64'
    }
)

demographic_df = dd.read_csv(
    demographic_file_path,
    dtype={'HOUSEHOLD_SIZE_DESC': 'object'}
)

product_df = dd.read_csv(product_file_path)

causal_df = dd.read_csv(
    causal_data_file_path,
    dtype={'display': 'object'}
)

coupon_df = dd.read_csv(coupon_file_path)

coupon_redemption_df = dd.read_csv(coupon_redemption_file_path)

campaign_df = dd.read_csv(campaign_file_path)

campaign_description_df = dd.read_csv(campaign_desc_file_path)


# Explore Each Dataset: First 5 Rows & Dataset Dimensions

## Transaction Dataset

In [ ]:
# Display first 5 rows of the dataset
transaction_df.head(5)

household_key    BASKET_ID  DAY  PRODUCT_ID  QUANTITY  SALES_VALUE  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   
2           2375  26984851472    1     1036325         1         0.99   
3           2375  26984851472    1     1082185         1         1.21   
4           2375  26984851472    1     8160430         1         1.50   

   STORE_ID  RETAIL_DISC  TRANS_TIME  WEEK_NO  COUPON_DISC  COUPON_MATCH_DISC  
0       364        -0.60        1631        1          0.0                0.0  
1       364         0.00        1631        1          0.0                0.0  
2       364        -0.30        1631        1          0.0                0.0  
3       364         0.00        1631        1          0.0                0.0  
4       364        -0.39        1631        1          0.0                0.0

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_df = transaction_df.shape[0].compute()
num_cols_transaction_df = len(transaction_df.columns)

print("Number of rows:", num_rows_transaction_df)
print("Number of columns:", num_cols_transaction_df)

Number of rows: 2595732
Number of columns: 12


In [ ]:
# Convert all column names to lowercase
transaction_df.columns = transaction_df.columns.str.lower()

# Display first 5 rows after the conversion
transaction_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   
2           2375  26984851472    1     1036325         1         0.99   
3           2375  26984851472    1     1082185         1         1.21   
4           2375  26984851472    1     8160430         1         1.50   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0       364        -0.60        1631        1          0.0                0.0  
1       364         0.00        1631        1          0.0                0.0  
2       364        -0.30        1631        1          0.0                0.0  
3       364         0.00        1631        1          0.0                0.0  
4       364        -0.39        1631        1          0.0                0.0

## Demographic Dataset

In [ ]:
# Display first 5 rows of the dataset
demographic_df.head()

AGE_DESC MARITAL_STATUS_CODE INCOME_DESC HOMEOWNER_DESC      HH_COMP_DESC  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   
2    25-34                   U      25-34K        Unknown     2 Adults Kids   
3    25-34                   U      75-99K      Homeowner     2 Adults Kids   
4    45-54                   B      50-74K      Homeowner     Single Female   

  HOUSEHOLD_SIZE_DESC KID_CATEGORY_DESC  household_key  
0                   2      None/Unknown              1  
1                   2      None/Unknown              7  
2                   3                 1              8  
3                   4                 2             13  
4                   1      None/Unknown             16

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_demographic_df = demographic_df.shape[0].compute()
num_cols_demographic_df= len(demographic_df.columns)

print("Number of rows:", num_rows_demographic_df)
print("Number of columns:", num_cols_demographic_df)

Number of rows: 801
Number of columns: 8


In [ ]:
# Convert all column names to lowercase
demographic_df.columns = demographic_df.columns.str.lower()

# Display first 5 rows after the conversion
demographic_df.head()

age_desc marital_status_code income_desc homeowner_desc      hh_comp_desc  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   
2    25-34                   U      25-34K        Unknown     2 Adults Kids   
3    25-34                   U      75-99K      Homeowner     2 Adults Kids   
4    45-54                   B      50-74K      Homeowner     Single Female   

  household_size_desc kid_category_desc  household_key  
0                   2      None/Unknown              1  
1                   2      None/Unknown              7  
2                   3                 1              8  
3                   4                 2             13  
4                   1      None/Unknown             16

## Product Dataset

In [ ]:
# Display first 5 rows of the dataset
product_df.head()

PRODUCT_ID  MANUFACTURER    DEPARTMENT     BRAND            COMMODITY_DESC  \
0       25671             2       GROCERY  National                  FRZN ICE   
1       26081             2  MISC. TRANS.  National  NO COMMODITY DESCRIPTION   
2       26093            69        PASTRY   Private                     BREAD   
3       26190            69       GROCERY   Private      FRUIT - SHELF STABLE   
4       26355            69       GROCERY   Private             COOKIES/CONES   

            SUB_COMMODITY_DESC CURR_SIZE_OF_PRODUCT  
0          ICE - CRUSHED/CUBED                22 LB  
1  NO SUBCOMMODITY DESCRIPTION                       
2         BREAD:ITALIAN/FRENCH                       
3                  APPLE SAUCE                50 OZ  
4            SPECIALTY COOKIES                14 OZ

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_product_df = product_df.shape[0].compute()
num_cols_product_df = len(product_df.columns)

print("Number of rows:", num_rows_product_df)
print("Number of columns:", num_cols_product_df)

Number of rows: 92353
Number of columns: 7


In [ ]:
# Convert all column names to lowercase
product_df.columns = product_df.columns.str.lower()

# Display first 5 rows after the conversion
product_df.head()

product_id  manufacturer    department     brand            commodity_desc  \
0       25671             2       GROCERY  National                  FRZN ICE   
1       26081             2  MISC. TRANS.  National  NO COMMODITY DESCRIPTION   
2       26093            69        PASTRY   Private                     BREAD   
3       26190            69       GROCERY   Private      FRUIT - SHELF STABLE   
4       26355            69       GROCERY   Private             COOKIES/CONES   

            sub_commodity_desc curr_size_of_product  
0          ICE - CRUSHED/CUBED                22 LB  
1  NO SUBCOMMODITY DESCRIPTION                       
2         BREAD:ITALIAN/FRENCH                       
3                  APPLE SAUCE                50 OZ  
4            SPECIALTY COOKIES                14 OZ

## Causal Dataset

In [ ]:
# Display first 5 rows of the dataset
causal_df.head()

PRODUCT_ID  STORE_ID  WEEK_NO display mailer
0       26190       286       70       0      A
1       26190       288       70       0      A
2       26190       289       70       0      A
3       26190       292       70       0      A
4       26190       293       70       0      A

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_causal_df = causal_df.shape[0].compute()
num_cols_causal_df = len(causal_df.columns)

print("Number of rows:", num_rows_causal_df)
print("Number of columns:", num_cols_causal_df)

Number of rows: 36786524
Number of columns: 5


In [ ]:
# Convert all column names to lowercase
causal_df.columns = causal_df.columns.str.lower()

# Display first 5 rows after the conversion
causal_df.head()

product_id  store_id  week_no display mailer
0       26190       286       70       0      A
1       26190       288       70       0      A
2       26190       289       70       0      A
3       26190       292       70       0      A
4       26190       293       70       0      A

## Coupon Dataset

In [ ]:
# Display first 5 rows of the dataset
coupon_df.head()

COUPON_UPC  PRODUCT_ID  CAMPAIGN
0  10000089061       27160         4
1  10000089064       27754         9
2  10000089073       28897        12
3  51800009050       28919        28
4  52100000076       28929        25

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_coupon_df = coupon_df.shape[0].compute()
num_cols_coupon_df = len(coupon_df.columns)

print("Number of rows:", num_rows_coupon_df)
print("Number of columns:", num_cols_coupon_df)

Number of rows: 124548
Number of columns: 3


In [ ]:
# Convert all column names to lowercase
coupon_df.columns = coupon_df.columns.str.lower()

# Display first 5 rows after the conversion
coupon_df.head()

coupon_upc  product_id  campaign
0  10000089061       27160         4
1  10000089064       27754         9
2  10000089073       28897        12
3  51800009050       28919        28
4  52100000076       28929        25

## Coupon Redemption Dataset

In [ ]:
# Display first 5 rows of the dataset
coupon_redemption_df.head()

household_key  DAY   COUPON_UPC  CAMPAIGN
0              1  421  10000085364         8
1              1  421  51700010076         8
2              1  427  54200000033         8
3              1  597  10000085476        18
4              1  597  54200029176        18

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_coupon_redemption_df = coupon_redemption_df.shape[0].compute()
num_cols_coupon_redemption_df = len(coupon_redemption_df.columns)

print("Number of rows:", num_rows_coupon_redemption_df)
print("Number of columns:", num_cols_coupon_redemption_df)

Number of rows: 2318
Number of columns: 4


In [ ]:
# Convert all column names to lowercase
coupon_redemption_df.columns = coupon_redemption_df.columns.str.lower()

# Display first 5 rows after the conversion
coupon_redemption_df.head()

household_key  day   coupon_upc  campaign
0              1  421  10000085364         8
1              1  421  51700010076         8
2              1  427  54200000033         8
3              1  597  10000085476        18
4              1  597  54200029176        18

## Campaign Dataset

In [ ]:
# Display first 5 rows of the dataset
campaign_df.head()

DESCRIPTION  household_key  CAMPAIGN
0       TypeA             17        26
1       TypeA             27        26
2       TypeA            212        26
3       TypeA            208        26
4       TypeA            192        26

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_campaign_df = campaign_df.shape[0].compute()
num_cols_campaign_df = len(campaign_df.columns)

print("Number of rows:", num_rows_campaign_df)
print("Number of columns:", num_cols_campaign_df)

Number of rows: 7208
Number of columns: 3


In [ ]:
# Convert all column names to lowercase
campaign_df.columns = campaign_df.columns.str.lower()

# Display first 5 rows after the conversion
campaign_df.head()

description  household_key  campaign
0       TypeA             17        26
1       TypeA             27        26
2       TypeA            212        26
3       TypeA            208        26
4       TypeA            192        26

## Campaign Description Dataset

In [ ]:
# Display first 5 rows of the dataset
campaign_description_df.head()

DESCRIPTION  CAMPAIGN  START_DAY  END_DAY
0       TypeB        24        659      719
1       TypeC        15        547      708
2       TypeB        25        659      691
3       TypeC        20        615      685
4       TypeB        23        646      684

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_campaign_description_df = campaign_description_df.shape[0].compute()
num_cols_campaign_description_df = len(campaign_description_df.columns)

print("Number of rows:", num_rows_campaign_description_df)
print("Number of columns:", num_cols_campaign_description_df)

Number of rows: 30
Number of columns: 4


In [ ]:
# Convert all column names to lowercase
campaign_description_df.columns = campaign_description_df.columns.str.lower()

# Display first 5 rows after the conversion
campaign_description_df.head()

description  campaign  start_day  end_day
0       TypeB        24        659      719
1       TypeC        15        547      708
2       TypeB        25        659      691
3       TypeC        20        615      685
4       TypeB        23        646      684

# Join Datasets

## Join Campaign Dataset With Campaign Description Dataset

In [ ]:
# Create a new column of 'day_range_tuple' of start day and end day for each row.
campaign_description_df['day_range_tuple'] = campaign_description_df.apply(
    lambda row: (row['start_day'], row['end_day']),
    axis=1,
    meta=('day_range_tuple', 'object')
)

In [ ]:
# Display first 5 rows after creating the new column
campaign_description_df.head()

description  campaign  start_day  end_day day_range_tuple
0       TypeB        24        659      719      (659, 719)
1       TypeC        15        547      708      (547, 708)
2       TypeB        25        659      691      (659, 691)
3       TypeC        20        615      685      (615, 685)
4       TypeB        23        646      684      (646, 684)

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_campaign_description_df = campaign_description_df.shape[0].compute()
num_cols_campaign_description_df = len(campaign_description_df.columns)

print("Number of rows:", num_rows_campaign_description_df)
print("Number of columns:", num_cols_campaign_description_df)

Number of rows: 30
Number of columns: 5


In [ ]:
# Join 'campaign_df' with 'campaign_description_df' on campaign
merged_campaign_df = campaign_df.merge(campaign_description_df, on='campaign', how='left')
merged_campaign_df.head(5)

description_x  household_key  campaign description_y  start_day  end_day  \
0         TypeA             17        26         TypeA        224      264   
1         TypeA             27        26         TypeA        224      264   
2         TypeA            212        26         TypeA        224      264   
3         TypeA            208        26         TypeA        224      264   
4         TypeA            192        26         TypeA        224      264   

  day_range_tuple  
0      (224, 264)  
1      (224, 264)  
2      (224, 264)  
3      (224, 264)  
4      (224, 264)

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_merged_campaign_df = merged_campaign_df.shape[0].compute()
num_cols_merged_campaign_df = len(merged_campaign_df.columns)

print("Number of rows:", num_rows_merged_campaign_df)
print("Number of columns:", num_cols_merged_campaign_df)

Number of rows: 7208
Number of columns: 7


## Join Transaction Dataset With Merged Dataset (Coupon & Coupon Description) & & Create a New Column 'received_campaign'

### Join Transaction Dataset With Merged Dataset (Coupon & Coupon Description)

In [ ]:
# Join 'campaign_df' with 'campaign_description_df' on campaign
transaction_campaign_df = transaction_df.merge(merged_campaign_df, on='household_key', how='left')

# Persist in memory
transaction_campaign_df = transaction_campaign_df.persist()

In [ ]:
# Display first 5 rows of the merged dataset
transaction_campaign_df.head(5)

household_key    basket_id  day  product_id  quantity  sales_value  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1004906         1         1.39   
2           2375  26984851472    1     1033142         1         0.82   
3           2375  26984851472    1     1033142         1         0.82   
4           2375  26984851472    1     1036325         1         0.99   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0       364         -0.6        1631        1          0.0                0.0   
1       364         -0.6        1631        1          0.0                0.0   
2       364          0.0        1631        1          0.0                0.0   
3       364          0.0        1631        1          0.0                0.0   
4       364         -0.3        1631        1          0.0                0.0   

  description_x  campaign description_y  start_day  end_day day_range_tuple  
0         TypeA      13.0         TypeA      504.0    551.0      (504, 551)  
1         TypeA      18.0         TypeA      587.0    642.0      (587, 642)  
2         TypeA      13.0         TypeA      504.0    551.0      (504, 551)  
3         TypeA      18.0         TypeA      587.0    642.0      (587, 642)  
4         TypeA      13.0         TypeA      504.0    551.0      (504, 551)

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_campaign_df = transaction_campaign_df.shape[0].compute()
num_cols_transaction_campaign_df= len(transaction_campaign_df.columns)

print("Number of rows:", num_rows_transaction_campaign_df)
print("Number of columns:", num_cols_transaction_campaign_df)

Number of rows: 14329057
Number of columns: 18


### Create a New Column 'received_campaign' in Merged Dataset (Transaction, Coupon & Description)

In [ ]:
# Extract the start day from the 'day_range_tuple' column.
transaction_campaign_start_days = transaction_campaign_df['day_range_tuple'].map(
    lambda x: x[0] if isinstance(x, tuple) else np.nan,
    meta=('start_day', 'float64')
)

# Extract the end day from the 'day_range_tuple' column.
transaction_campaign_end_days = transaction_campaign_df['day_range_tuple'].map(
    lambda x: x[1] if isinstance(x, tuple) else np.nan,
    meta=('end_day', 'float64')
)

# Create a new column 'received_campaign' to indicate if 'day' in trasanction dataset falls within the campaign period (Yes or No)
transaction_campaign_df['received_campaign'] = (
    (transaction_campaign_df['day'] >= transaction_campaign_start_days) &
    (transaction_campaign_df['day'] <= transaction_campaign_end_days)
).map({True: 'yes', False: 'no'}, meta=('in_campaign', 'object'))

# Persist in memory
transaction_campaign_df = transaction_campaign_df.persist()

# Display first 5 rows after creating the new column
transaction_campaign_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1004906         1         1.39   
2           2375  26984851472    1     1033142         1         0.82   
3           2375  26984851472    1     1033142         1         0.82   
4           2375  26984851472    1     1036325         1         0.99   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0       364         -0.6        1631        1          0.0                0.0   
1       364         -0.6        1631        1          0.0                0.0   
2       364          0.0        1631        1          0.0                0.0   
3       364          0.0        1631        1          0.0                0.0   
4       364         -0.3        1631        1          0.0                0.0   

  description_x  campaign description_y  start_day  end_day day_range_tuple  \
0         TypeA      13.0         TypeA      504.0    551.0      (504, 551)   
1         TypeA      18.0         TypeA      587.0    642.0      (587, 642)   
2         TypeA      13.0         TypeA      504.0    551.0      (504, 551)   
3         TypeA      18.0         TypeA      587.0    642.0      (587, 642)   
4         TypeA      13.0         TypeA      504.0    551.0      (504, 551)   

  received_campaign  
0                no  
1                no  
2                no  
3                no  
4                no

In [ ]:
# Show rows where 'received_campaign' is 'yes'
transaction_campaign_yes_df = transaction_campaign_df [transaction_campaign_df['received_campaign'] == 'yes']

# View first 20 rows
transaction_campaign_yes_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
3355440           2223  30175934369  224      834103         1         1.29   
3355443           2223  30175934369  224      973135         1         2.99   
3355446           2223  30175934369  224     1011300         1         1.29   
3355449           2223  30175934369  224     1110572         1         1.29   
3367072           1366  30190041499  224     1051245         2         3.78   

         store_id  retail_disc  trans_time  week_no  coupon_disc  \
3355440       381          0.0           2       33          0.0   
3355443       381         -0.5           2       33          0.0   
3355446       381          0.0           2       33          0.0   
3355449       381          0.0           2       33          0.0   
3367072     31862          0.0        1559       33          0.0   

         coupon_match_disc description_x  campaign description_y  start_day  \
3355440                0.0         TypeA      26.0         TypeA      224.0   
3355443                0.0         TypeA      26.0         TypeA      224.0   
3355446                0.0         TypeA      26.0         TypeA      224.0   
3355449                0.0         TypeA      26.0         TypeA      224.0   
3367072                0.0         TypeA      26.0         TypeA      224.0   

         end_day day_range_tuple received_campaign  
3355440    264.0      (224, 264)               yes  
3355443    264.0      (224, 264)               yes  
3355446    264.0      (224, 264)               yes  
3355449    264.0      (224, 264)               yes  
3367072    264.0      (224, 264)               yes

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_campaign_yes_df = transaction_campaign_yes_df.shape[0].compute()
num_cols_transaction_campaign_yes_df = len(transaction_campaign_yes_df.columns)

print("Number of rows in 'transaction_campaign_yes_df' that show yes in 'received_campaign' column:", num_rows_transaction_campaign_yes_df)
print("Number of columns in 'transaction_campaign_yes_df' that show yes in 'received_campaign' column:", num_cols_transaction_campaign_yes_df)

Number of rows in 'transaction_campaign_yes_df' that show yes in 'received_campaign' column: 994137
Number of columns in 'transaction_campaign_yes_df' that show yes in 'received_campaign' column: 19


### Join Transaction Dataset With 'transaction_campaign_yes' Dataset (Where 'received_campaign' is yes)

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_df = transaction_df.shape[0].compute()
num_cols_transaction_df = len(transaction_df.columns)

print("Number of rows in 'transaction_df':", num_rows_transaction_df)
print("Number of columns in 'transaction_df':", num_cols_transaction_df)

Number of rows in 'transaction_df': 2595732
Number of columns in 'transaction_df': 12


In [ ]:
# Join 'transaction_df' with 'transaction_campaign_df' on household key, product_id and day
transaction_transaction_campaign_yes_df = transaction_df.merge(
    transaction_campaign_yes_df,
    on=['household_key', 'product_id', 'day'],
    how='left'
)

In [ ]:
# Display first 5 rows of the merged dataset
transaction_transaction_campaign_yes_df.head()

household_key  basket_id_x  day  product_id  quantity_x  sales_value_x  \
0           1453  32811562421  392      986912           1           1.99   
1            459  32811565816  392      885290           1           1.50   
2            608  32811570135  392     1096573           1           3.99   
3            878  32811571492  392      861819           1           1.00   
4            878  32811571492  392      866227           1           0.50   

   store_id_x  retail_disc_x  trans_time_x  week_no_x  ...  week_no_y  \
0       31862           0.00          2059         57  ...        NaN   
1       31762          -0.99          1553         57  ...        NaN   
2         297          -0.50           958         57  ...       57.0   
3         297          -0.65          2002         57  ...       57.0   
4         297           0.00          2002         57  ...       57.0   

   coupon_disc_y  coupon_match_disc_y  description_x  campaign  description_y  \
0            NaN                  NaN           <NA>       NaN           <NA>   
1            NaN                  NaN           <NA>       NaN           <NA>   
2            0.0                  0.0          TypeB       5.0          TypeB   
3            0.0                  0.0          TypeB       5.0          TypeB   
4            0.0                  0.0          TypeB       5.0          TypeB   

   start_day  end_day  day_range_tuple  received_campaign  
0        NaN      NaN              NaN                NaN  
1        NaN      NaN              NaN                NaN  
2      377.0    411.0       [377, 411]                yes  
3      377.0    411.0       [377, 411]                yes  
4      377.0    411.0       [377, 411]                yes  

[5 rows x 28 columns]

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame after merging
num_rows_transaction_transaction_campaign_yes_df = transaction_transaction_campaign_yes_df.shape[0].compute()
num_cols_transaction_transaction_campaign_yes_df = len(transaction_transaction_campaign_yes_df.columns)

print("Number of rows:", num_rows_transaction_transaction_campaign_yes_df)
print("Number of columns:", num_cols_transaction_transaction_campaign_yes_df)

Number of rows: 2828875
Number of columns: 28


In [ ]:
# Select columns that do not end with "_y"
cols_to_keep_not_y = [col for col in transaction_transaction_campaign_yes_df.columns
                if not col.endswith('_y')]

# Keep only columns that do not end with "_y"
transaction_transaction_campaign_yes_df = transaction_transaction_campaign_yes_df[cols_to_keep_not_y]

# Drop the 'day_range_tuple' column
transaction_transaction_campaign_yes_df = transaction_transaction_campaign_yes_df.drop(columns='day_range_tuple')

In [ ]:
# Rename columns to remove the '_x' suffix added during merge
transaction_transaction_campaign_yes_df = (
    transaction_transaction_campaign_yes_df
    .rename(columns={col: col[:-2] for col in transaction_transaction_campaign_yes_df.columns
                     if col.endswith('_x')})
)

In [ ]:
transaction_transaction_campaign_yes_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           1453  32811562421  392      986912         1         1.99   
1            459  32811565816  392      885290         1         1.50   
2            608  32811570135  392     1096573         1         3.99   
3            878  32811571492  392      861819         1         1.00   
4            878  32811571492  392      866227         1         0.50   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0     31862         0.00        2059       57          0.0                0.0   
1     31762        -0.99        1553       57          0.0                0.0   
2       297        -0.50         958       57          0.0                0.0   
3       297        -0.65        2002       57          0.0                0.0   
4       297         0.00        2002       57          0.0                0.0   

  description  campaign  start_day  end_day received_campaign  
0        <NA>       NaN        NaN      NaN               NaN  
1        <NA>       NaN        NaN      NaN               NaN  
2       TypeB       5.0      377.0    411.0               yes  
3       TypeB       5.0      377.0    411.0               yes  
4       TypeB       5.0      377.0    411.0               yes

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame after merging
num_rows_transaction_transaction_campaign_yes_df = transaction_transaction_campaign_yes_df.shape[0].compute()
num_cols_transaction_transaction_campaign_yes_df = len(transaction_transaction_campaign_yes_df.columns)

print("Number of rows:", num_rows_transaction_transaction_campaign_yes_df)
print("Number of columns:", num_cols_transaction_transaction_campaign_yes_df)

Number of rows: 2828875
Number of columns: 17


In [ ]:
# List all columns except 'campaign', 'description', 'start_day' and 'end_day'
cols_to_check = [col for col in transaction_transaction_campaign_yes_df.columns
                 if col not in ['campaign', 'description', 'start_day', 'end_day']]

# Create a boolean column marking duplicates within each partition
def mark_duplicates(df, subset):
    return df[df.duplicated(subset=subset, keep=False)]

# Apply to each partition
duplicates_transaction_transaction_campaign_yes_df = transaction_transaction_campaign_yes_df.map_partitions(
    mark_duplicates, subset=cols_to_check
)

# Compute number of duplicate rows
num_duplicates_transaction_transaction_campaign_yes_df = len(duplicates_transaction_transaction_campaign_yes_df.compute())

print(f"Number of duplicate rows: {num_duplicates_transaction_transaction_campaign_yes_df}")

duplicates_transaction_transaction_campaign_yes_df = duplicates_transaction_transaction_campaign_yes_df.persist()

Number of duplicate rows: 431528


In [ ]:
# Remove duplicates based on the subset of columns
transaction_campaign_no_duplicates_df = transaction_transaction_campaign_yes_df.drop_duplicates(
    subset=cols_to_check,
    keep='first'
)

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_campaign_no_duplicates_df = transaction_campaign_no_duplicates_df.shape[0].compute()
num_cols_transaction_campaign_no_duplicates_df = len(transaction_campaign_no_duplicates_df.columns)

print("Number of rows:", num_rows_transaction_campaign_no_duplicates_df)
print("Number of columns:", num_cols_transaction_campaign_no_duplicates_df)

Number of rows: 2595732
Number of columns: 17


In [ ]:
# Fill missing values in 'received_campaign' column with 'no'
transaction_campaign_no_duplicates_df['received_campaign'] = (
    transaction_campaign_no_duplicates_df['received_campaign']
    .fillna('no')
)

In [ ]:
# Display first 5 rows of the dataset
transaction_campaign_no_duplicates_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           1453  32811562421  392      986912         1         1.99   
1            459  32811565816  392      885290         1         1.50   
2            608  32811570135  392     1096573         1         3.99   
3            878  32811571492  392      861819         1         1.00   
4            878  32811571492  392      866227         1         0.50   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0     31862         0.00        2059       57          0.0                0.0   
1     31762        -0.99        1553       57          0.0                0.0   
2       297        -0.50         958       57          0.0                0.0   
3       297        -0.65        2002       57          0.0                0.0   
4       297         0.00        2002       57          0.0                0.0   

  description  campaign  start_day  end_day received_campaign  
0        <NA>       NaN        NaN      NaN                no  
1        <NA>       NaN        NaN      NaN                no  
2       TypeB       5.0      377.0    411.0               yes  
3       TypeB       5.0      377.0    411.0               yes  
4       TypeB       5.0      377.0    411.0               yes

## Join Merged Dataset (Transaction, Coupon & Coupon Description) With Demographic Dataset

In [ ]:
# Join 'transaction_df' with 'demographic_df' on household_key
transaction_demographic_df = transaction_campaign_no_duplicates_df.merge(demographic_df, on='household_key', how='left')

# Show all columns
pd.set_option('display.max_columns', None)

# Persist in memory
transaction_demographic_df = transaction_demographic_df.persist()

# Display first 5 rows of the merged dataset
transaction_demographic_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           1453  32811562421  392      986912         1         1.99   
1            459  32811565816  392      885290         1         1.50   
2            608  32811570135  392     1096573         1         3.99   
3            878  32811571492  392      861819         1         1.00   
4            878  32811571492  392      866227         1         0.50   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0     31862         0.00        2059       57          0.0                0.0   
1     31762        -0.99        1553       57          0.0                0.0   
2       297        -0.50         958       57          0.0                0.0   
3       297        -0.65        2002       57          0.0                0.0   
4       297         0.00        2002       57          0.0                0.0   

  description  campaign  start_day  end_day received_campaign age_desc  \
0        <NA>       NaN        NaN      NaN                no    45-54   
1        <NA>       NaN        NaN      NaN                no     <NA>   
2       TypeB       5.0      377.0    411.0               yes    25-34   
3       TypeB       5.0      377.0    411.0               yes    25-34   
4       TypeB       5.0      377.0    411.0               yes    25-34   

  marital_status_code income_desc homeowner_desc   hh_comp_desc  \
0                   A    125-149K      Homeowner  2 Adults Kids   
1                <NA>        <NA>           <NA>           <NA>   
2                   B      25-34K         Renter   1 Adult Kids   
3                   U      35-49K        Unknown   1 Adult Kids   
4                   U      35-49K        Unknown   1 Adult Kids   

  household_size_desc kid_category_desc  
0                   3                 1  
1                <NA>              <NA>  
2                  5+                3+  
3                   3                 2  
4                   3                 2

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_df = transaction_demographic_df.shape[0].compute()
num_cols_transaction_demographic_df = len(transaction_demographic_df.columns)

print("Number of rows:", num_rows_transaction_demographic_df)
print("Number of columns:", num_cols_transaction_demographic_df)

Number of rows: 2595732
Number of columns: 24


## Join Merged Dataset (Transaction Coupon, Coupon Description & Demographic) With Product Dataset

In [ ]:
# Join 'transaction_demographic_df' with 'product_df' on product_id
transaction_demographic_product_df = transaction_demographic_df.merge(product_df, on='product_id', how='left')

# Show all columns
pd.set_option('display.max_columns', None)

# Persist in memory
transaction_demographic_product_df = transaction_demographic_product_df.persist()

# Display first 5 rows of the merged dataset
transaction_demographic_product_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           1453  32811562421  392      986912         1         1.99   
1            459  32811565816  392      885290         1         1.50   
2            608  32811570135  392     1096573         1         3.99   
3            878  32811571492  392      861819         1         1.00   
4            878  32811571492  392      866227         1         0.50   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0     31862         0.00        2059       57          0.0                0.0   
1     31762        -0.99        1553       57          0.0                0.0   
2       297        -0.50         958       57          0.0                0.0   
3       297        -0.65        2002       57          0.0                0.0   
4       297         0.00        2002       57          0.0                0.0   

  description  campaign  start_day  end_day received_campaign age_desc  \
0        <NA>       NaN        NaN      NaN                no    45-54   
1        <NA>       NaN        NaN      NaN                no     <NA>   
2       TypeB       5.0      377.0    411.0               yes    25-34   
3       TypeB       5.0      377.0    411.0               yes    25-34   
4       TypeB       5.0      377.0    411.0               yes    25-34   

  marital_status_code income_desc homeowner_desc   hh_comp_desc  \
0                   A    125-149K      Homeowner  2 Adults Kids   
1                <NA>        <NA>           <NA>           <NA>   
2                   B      25-34K         Renter   1 Adult Kids   
3                   U      35-49K        Unknown   1 Adult Kids   
4                   U      35-49K        Unknown   1 Adult Kids   

  household_size_desc kid_category_desc  manufacturer department     brand  \
0                   3                 1          4166       DELI  National   
1                <NA>              <NA>           910    GROCERY  National   
2                  5+                3+           608    GROCERY  National   
3                   3                 2            69    GROCERY   Private   
4                   3                 2             2     PASTRY  National   

                   commodity_desc              sub_commodity_desc  \
0                   PREPARED FOOD      PREP FD: SIDE DISHES (HOT)   
1          BAKED BREAD/BUNS/ROLLS  MAINSTREAM WHEAT/MULTIGRAIN BR   
2          FRZN MEAT/MEAT DINNERS               SNACKS/APPETIZERS   
3  CONVENIENT BRKFST/WHLSM SNACKS                    FRUIT SNACKS   
4                BREAKFAST SWEETS                   SW GDS:DONUTS   

  curr_size_of_product  
0                       
1                20 OZ  
2                16 OZ  
3               5.4 OZ  
4

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_df = transaction_demographic_product_df.shape[0].compute()
num_cols_transaction_demographic_product_df = len(transaction_demographic_product_df.columns)

print("Number of rows:", num_rows_transaction_demographic_product_df)
print("Number of columns:", num_cols_transaction_demographic_product_df)

Number of rows: 2595732
Number of columns: 30


## Join Merged Dataset (Transaction, Coupon, Coupon Description, Demographic & Product) With Causal Dataset

In [ ]:
# Join 'transaction_demographic_product_df' with 'causal_df' on product_id, store_id and week_no
transaction_demographic_product_causal_df = transaction_demographic_product_df.merge(causal_df, on=['product_id', 'store_id', 'week_no'], how='left')

# Persist in memory
transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.persist()

In [ ]:
# Show all columns
pd.set_option('display.max_columns', None)

# Display first 5 rows of the merged dataset
transaction_demographic_product_causal_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0            459  32811565816  392     1039682         1         0.99   
1            330  32811585245  392      994928         1         0.74   
2            276  32811586367  392      847232         1         1.49   
3            276  32811586367  392     1112467         1         3.45   
4            844  32811591209  392     8011268         1         3.49   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0     31762          0.0        1553       57          0.0                0.0   
1       362          0.0        1112       57          0.0                0.0   
2       362          0.0        1912       57          0.0                0.0   
3       362          0.0        1912       57          0.0                0.0   
4       317          0.0        1459       57          0.0                0.0   

  description  campaign  start_day  end_day received_campaign age_desc  \
0        <NA>       NaN        NaN      NaN                no     <NA>   
1        <NA>       NaN        NaN      NaN                no     <NA>   
2        <NA>       NaN        NaN      NaN                no    35-44   
3        <NA>       NaN        NaN      NaN                no    35-44   
4        <NA>       NaN        NaN      NaN                no    45-54   

  marital_status_code income_desc homeowner_desc      hh_comp_desc  \
0                <NA>        <NA>           <NA>              <NA>   
1                <NA>        <NA>           <NA>              <NA>   
2                   U      35-49K        Unknown       Single Male   
3                   U      35-49K        Unknown       Single Male   
4                   A    200-249K      Homeowner  2 Adults No Kids   

  household_size_desc kid_category_desc  manufacturer department     brand  \
0                <NA>              <NA>             2    PRODUCE  National   
1                <NA>              <NA>            69    GROCERY   Private   
2                   1      None/Unknown           869    GROCERY  National   
3                   1      None/Unknown          2872       MEAT  National   
4                   2      None/Unknown            69       DELI   Private   

     commodity_desc         sub_commodity_desc curr_size_of_product display  \
0  VEGETABLES SALAD            VARIETY LETTUCE                24 CT    <NA>   
1              EGGS              EGGS - MEDIUM                 1 DZ    <NA>   
2              SOUP                   DRY SOUP           4.2 OZ/2CT    <NA>   
3              BEEF                CHOICE BEEF                         <NA>   
4           CHEESES  CHEESE: NATURAL PREPORTND                 8 OZ    <NA>   

  mailer  
0   <NA>  
1   <NA>  
2   <NA>  
3   <NA>  
4   <NA>

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.shape[0].compute()
num_cols_transaction_demographic_product_causal_df = len(transaction_demographic_product_causal_df.columns)

print("Number of rows in transaction_demographic_product_causal_df:", num_rows_transaction_demographic_product_causal_df)
print("Number of columns in transaction_demographic_product_causal_df:", num_cols_transaction_demographic_product_causal_df)

Number of rows in transaction_demographic_product_causal_df: 2596590
Number of columns in transaction_demographic_product_causal_df: 32


In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_df = transaction_demographic_product_df.shape[0].compute()
num_cols_transaction_demographic_product_df = len(transaction_demographic_product_df.columns)

print("Number of rows in transaction_demographic_product_df:", num_rows_transaction_demographic_product_df)
print("Number of columns in transaction_demographic_product_df:", num_cols_transaction_demographic_product_df)

Number of rows in transaction_demographic_product_df: 2595732
Number of columns in transaction_demographic_product_df: 30


In [ ]:
# Calculate the number of row difference between 'transaction_demographic_product_causal_df' and 'transaction_demographic_product_df'
row_difference_1 = num_rows_transaction_demographic_product_causal_df - num_rows_transaction_demographic_product_df

print(f"Row difference between 'transaction_demographic_product_causal_df' and 'transaction_demographic_product_df': {row_difference_1} rows")

Row difference between 'transaction_demographic_product_causal_df' and 'transaction_demographic_product_df': 858 rows


In [ ]:
# List all columns except 'display' and 'mailer'
cols_to_check = [col for col in transaction_demographic_product_causal_df.columns
                 if col not in ['display', 'mailer']]

# Create a boolean column marking duplicates within each partition
def mark_duplicates(df, subset):
    return df[df.duplicated(subset=subset, keep=False)]

# Apply to each partition
duplicates_transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.map_partitions(
    mark_duplicates, subset=cols_to_check
)

# Compute number of duplicate rows
num_duplicates_transaction_demographic_product_causal_df  = len(duplicates_transaction_demographic_product_causal_df.compute())

print(f"Number of duplicate rows: {num_duplicates_transaction_demographic_product_causal_df}")

Number of duplicate rows: 1716


In [ ]:
# Display first 5 rows of the merged dataset
duplicates_transaction_demographic_product_causal_df.head(5)

# After the left join, the the number of rows increased from 2,595,732 to 2,596,590 (858 rows).
# This is due to multiple matching rows for 'display' and 'mailer' columns in the right table (causal_df).

household_key    basket_id  day  product_id  quantity  sales_value  \
2396             17  32989921691  405      994928         6          2.0   
2397             17  32989921691  405      994928         6          2.0   
2793           1208  33015899361  407      994928         3          1.0   
2794           1208  33015899361  407      994928         3          1.0   
3708           1403  33065585590  413     5569230         1          2.5   

      store_id  retail_disc  trans_time  week_no  coupon_disc  \
2396       361        -2.44        1356       59          0.0   
2397       361        -2.44        1356       59          0.0   
2793       361        -1.22        2254       59          0.0   
2794       361        -1.22        2254       59          0.0   
3708       406        -2.19          47       60          0.0   

      coupon_match_disc description  campaign  start_day  end_day  \
2396                0.0        <NA>       NaN        NaN      NaN   
2397                0.0        <NA>       NaN        NaN      NaN   
2793                0.0        <NA>       NaN        NaN      NaN   
2794                0.0        <NA>       NaN        NaN      NaN   
3708                0.0       TypeA       8.0      412.0    460.0   

     received_campaign age_desc marital_status_code income_desc  \
2396                no      65+                   B   Under 15K   
2397                no      65+                   B   Under 15K   
2793                no     <NA>                <NA>        <NA>   
2794                no     <NA>                <NA>        <NA>   
3708               yes     <NA>                <NA>        <NA>   

     homeowner_desc      hh_comp_desc household_size_desc kid_category_desc  \
2396      Homeowner  2 Adults No Kids                   2      None/Unknown   
2397      Homeowner  2 Adults No Kids                   2      None/Unknown   
2793           <NA>              <NA>                <NA>              <NA>   
2794           <NA>              <NA>                <NA>              <NA>   
3708           <NA>              <NA>                <NA>              <NA>   

      manufacturer department     brand commodity_desc  \
2396            69    GROCERY   Private           EGGS   
2397            69    GROCERY   Private           EGGS   
2793            69    GROCERY   Private           EGGS   
2794            69    GROCERY   Private           EGGS   
3708          1208    GROCERY  National    SOFT DRINKS   

                  sub_commodity_desc curr_size_of_product display mailer  
2396                   EGGS - MEDIUM                 1 DZ       0      D  
2397                   EGGS - MEDIUM                 1 DZ       6      D  
2793                   EGGS - MEDIUM                 1 DZ       0      D  
2794                   EGGS - MEDIUM                 1 DZ       6      D  
3708  SOFT DRINKS 12/18&15PK CAN CAR                12 OZ       0      D

In [ ]:
# Count how many duplicate rows have display equal to "0" in 'duplicates_transaction_demographic_product_causal_df'
display_equal_0 = duplicates_transaction_demographic_product_causal_df[
    duplicates_transaction_demographic_product_causal_df['display'] == "0"
].shape[0].compute()

print("Number of duplicate rows have display equal to '0' in 'duplicates_transaction_demographic_product_causal_df':", display_equal_0, 'rows')

Number of duplicate rows have display equal to '0' in 'duplicates_transaction_demographic_product_causal_df': 858 rows


In [ ]:
# As our aim is to focus on transactional-level patterns for customer segmentation, the extra rows don't represent new purchases.
# They exist because a single transaction matched multiple display locations (the same product purchased in the same transaction may have different display locations.)
# Keeping these duplicates could inflate the number of transactions and bias our customer segmnetation, making it look like some customers buy more than they actually do.

# Remove all duplicate rows where display column equals to "0" (0 means not on display) in transaction_demographic_product_causal_df
# Convert Dask DataFrame to pandas
transaction_demographic_product_causal_df_pandas = transaction_demographic_product_causal_df.compute()

# Define columns to check for duplicates (excluding 'display' and 'mailer')
cols_to_identify_transaction_demographic_product_causal_df_pandas = [col for col in transaction_demographic_product_causal_df_pandas.columns if col not in ['display', 'mailer']]

# Remove duplicate rows where display column equals to "0" (0 means not on display) in transaction_demographic_product_causal_df
transaction_demographic_product_causal_df_pandas = transaction_demographic_product_causal_df_pandas[~(
    (transaction_demographic_product_causal_df_pandas.duplicated(subset=cols_to_identify_transaction_demographic_product_causal_df_pandas, keep=False)) &
    (transaction_demographic_product_causal_df_pandas['display'] == "0")
)]

# Convert back to Dask DataFrame
transaction_demographic_product_causal_df = dd.from_pandas(transaction_demographic_product_causal_df_pandas)

# Persist in memory
transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.persist()

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.shape[0].compute()
num_cols_transaction_demographic_product_causal_df = len(transaction_demographic_product_causal_df.columns)

print("Number of rows in transaction_demographic_product_causal_df:", num_rows_transaction_demographic_product_causal_df)
print("Number of columns in transaction_demographic_product_causal_df:", num_cols_transaction_demographic_product_causal_df)

Number of rows in transaction_demographic_product_causal_df: 2595732
Number of columns in transaction_demographic_product_causal_df: 32


## Join Coupon Dataset With Coupon Redemption Dataset

In [ ]:
# Join 'coupon_df' with 'coupon_redemption_df' on coupon_upc and campaign
merged_coupon_df = coupon_df.merge(coupon_redemption_df, on=['coupon_upc', 'campaign'], how='left')

# Persist in memory
merged_coupon_df = merged_coupon_df.persist()

In [ ]:
# Show all columns
pd.set_option('display.max_columns', None)

# Display first 5 rows of the merged dataset
merged_coupon_df.head(5)

coupon_upc  product_id  campaign  household_key    day
0  10000089061       27160         4            NaN    NaN
1  10000089064       27754         9          321.0  446.0
2  10000089064       27754         9         1773.0  439.0
3  10000089073       28897        12         1229.0  491.0
4  10000089073       28897        12         1633.0  497.0

In [ ]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_merged_coupon_df = merged_coupon_df.shape[0].compute()
num_cols_merged_coupon_df = len(merged_coupon_df.columns)

print("Number of rows:", num_rows_merged_coupon_df)
print("Number of columns:", num_cols_merged_coupon_df)

Number of rows: 2218551
Number of columns: 5


**Interpretation for merged_coupon_df:**

**Row 0:**
- Coupon 10000089061 for product 27160 in campaign 4.
- household_key = NaN, day = NaN → *coupon exists but has not been redeemed yet.*

**Row 1:**
- Coupon 10000089064 for product 27754 in campaign 9.
- household_key = 321, day = 446 → *household 321 redeemed this coupon on day 446.*

**Row 2:**
- Same coupon and campaign as row 1.
- household_key = 1773, day = 439 → *same coupon redeemed by another household 1773 on day 439.*

In [ ]:
# Save 'transaction_demographic_product_causal_df' and 'merged_coupon_df' as CSV file

transaction_demographic_product_causal_df.compute().to_csv('/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/2. Preprocessed Datasets/transaction_demographic_product_causal_df.csv', index=False)
merged_coupon_df.compute().to_csv('/content/drive/MyDrive/0. Colab Notebooks (DATA 60X Project)/merged_coupon_df.csv', index=False)